[JENKINS-54950] - use bindJSON(this,json) as default implementation for Descriptor#configure

ndeloof opened the pull request:
This change is intended to promote use of databinding relying on DataBoundSetters in Descriptors, as required by JCasC. 

See [JENKINS-54950](https://issues.jenkins-ci.org/browse/JENKINS-54950).

### Proposed changelog entries

* Internal: use bindJSON() as default implementation for Descriptor#configure

### Submitter checklist

- [x] JIRA issue is well described
- [x] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). 
- [ ] Appropriate autotests or explanation to why this change has no tests
- [n/a] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

ndeloof committed:
rely on default Descriptor.configure implementation
introduce @ManagedBy annotation to flag Jenkins' fields actually managed by a dedicated component

Signed-off-by: Nicolas De Loof <nicolas.deloof@gmail.com>

ndeloof committed:
demonstrate support for @DefaultValue in jelly tags

Signed-off-by: Nicolas De Loof <nicolas.deloof@gmail.com>

ndeloof committed:
use bindJSON(this,json) as default implementation for Descriptor#configure

Signed-off-by: Nicolas De Loof <nicolas.deloof@gmail.com>

jvz reviewed (approved):
The less code required for a plugin the better!

ndeloof committed:
Prevent bind and save() being executed twice

Signed-off-by: Nicolas De Loof <nicolas.deloof@gmail.com>

ndeloof committed:
rely on BulkChange so setters can invoke save()

Signed-off-by: Nicolas De Loof <nicolas.deloof@gmail.com>

daniel-beck commented:
So basically boilerplate code like https://github.com/jenkinsci/login-theme-plugin/blob/f9981980d1db118eaddae71f077df54e67589ac5/src/main/java/io/jenkins/plugins/logintheme/LoginTheme.java#L86...L91 will no longer be required. Nice.

daniel-beck commented:
I'd be more comfortable with tests confirming this works as intended though TBH.

ndeloof commented:
yes indeed. Make this the default, so people only override `configure` for some very specific corner cases, just like `newInstance` usually doesn't need any customisation by plugin maintainers

oleg-nenashev reviewed (approved):
LGTM. Some tests would be nice. Or you could just remove some obsolete code in Descriptor implementations within the core (a follow-up ticket for it would be nice)

ndeloof commented:
@oleg-nenashev I indeed plan to review Descriptor implementations in core.
Main point I'd like to get review here is how to handle calls to super.configure, with risks for save() to be called twice. @jglick already recommended to rely on `BulkChange` for this purpose

ndeloof committed:
Rely on BulkChange to avoid repeated calls to save()
introduce configureNested to avoid code duplication

Signed-off-by: Nicolas De Loof <nicolas.deloof@gmail.com>

ndeloof commented:
I searched for a candidate Descriptor in jenkins-core I could use to write a testcase, but all those I've checked do some custom computation or adapt to legacy data structure.

ndeloof committed:
no need to override configure in SCMTrigger.DescriptorImpl
setPollingThreadCount as public method has to invoke save()
it also already handle range check, no needs to duplicate in configure()

Signed-off-by: Nicolas De Loof <nicolas.deloof@gmail.com>

ndeloof commented:
@daniel-beck implemented a test case as hudson.model.DescriptorTest#descriptor_config_roudtrip
relying on (simplified) SCMTrigger.DescriptorImpl. 
As suggested by @oleg-nenashev some further descriptor cleanup could be proposed in a follow-up PR.

oleg-nenashev commented on core/src/main/java/hudson/model/Descriptor.java:
```diff
@@ -806,8 +808,31 @@ public boolean configure( StaplerRequest req ) throws FormException {
+    /**
+     * Configure descriptor instance form structured form submission where this specific descriptor is nested in
+     * a larger payload, which typically happens when Descriptor doesn't have a dedicated configuration view.
+     */
```
Add `@since TODO` if you want it to be a part of API. Otherwise makes sense to restrict it

Wadeck replied:
Change to consider 👍 as for all the new public methods in this PR

oleg-nenashev commented on core/src/main/java/hudson/model/Descriptor.java:
```diff
@@ -806,8 +808,31 @@ public boolean configure( StaplerRequest req ) throws FormException {
+            return configure(req);
+        }
+
+        try (BulkChange bc = new BulkChange(this)) {
```
IMO it should check `if (BulkChange.contains(this))` and invoke new BulkChange only when needed

ndeloof replied:
AFAIK BulkChange by design fully support being nested, doesn't it ?
BulkChange.contains is mostly used by `save` to bypass actual serialization on disk

ndeloof replied:
I can't find any other sample of using BulkChange with such an additional contains check

oleg-nenashev replied:
I should review it more carefully

ndeloof replied:
I also can't imagine any significant use-case to invoke `configure` within an existing BulkChange on the exact same object that would me this `contains` check relevant.

oleg-nenashev reviewed (dismissed):
Looks reasonable. IMO the transaction logic should be improved to run IFF there is no bending BulkChange for the object

timja commented on core/src/main/java/jenkins/tools/GlobalToolConfiguration.java:
```diff
@@ -24,6 +24,7 @@
 package jenkins.tools;
 
 import com.google.common.base.Predicate;
+import hudson.BulkChange;
```
BulkChange doesn't appear to be used here?

ndeloof replied:
indeed, fixed

daniel-beck commented on core/src/main/java/hudson/triggers/SCMTrigger.java:
```diff
@@ -351,21 +351,6 @@ public boolean isPollingThreadCountOptionVisible() {
-        public boolean configure(StaplerRequest req, JSONObject json) throws FormException {
-            String t = json.optString("pollingThreadCount",null);
-            if (doCheckPollingThreadCount(t).kind != FormValidation.Kind.OK) {
-                setPollingThreadCount(THREADS_DEFAULT);
```
Semantics change: This resets to default when invalid from user submission, the setter goes to the nearest boundary.

Also, don't we need a `@DataboundSetter` for this to work?

ndeloof replied:
Semantic here indeed was heterogeneous. descriptor didn't used the same strategy when configured by webUI and direct setter invocation. 
I'd personally would prefer it reject improper values vs silently prefer a sane one without caller to know his proposed value was rejected, but this goes beyond this PR.

ndeloof replied:
bindJSON(Object, json) do rely on javabean setters, no need for `@DataboundSetter` annotation - while this is required for Describable relying on bindJSON(class, json). Don't ask me about the logic here :)

daniel-beck commented on core/src/main/java/hudson/model/Descriptor.java:
```diff
@@ -814,21 +814,27 @@ public boolean configure( StaplerRequest req, JSONObject json ) throws FormExcep
+     * Configure descriptor instance form structured form submission where this specific descriptor is nested in
+     * a larger payload, which typically happens when Descriptor doesn't have a dedicated configuration view.
+     */
+    public boolean configureNested( StaplerRequest req, JSONObject json) throws FormException, IOException {
```
`protected`?

daniel-beck commented on core/src/main/java/hudson/model/Descriptor.java:
```diff
@@ -814,21 +814,27 @@ public boolean configure( StaplerRequest req, JSONObject json ) throws FormExcep
+    public boolean configureNested( StaplerRequest req, JSONObject json) throws FormException, IOException {
+        String name = getJsonSafeClassName();
+        JSONObject js = json.has(name) ? json.getJSONObject(name) : new JSONObject(); // if it doesn't have the property, the method returns invalid null object.
+        json.putAll(js);
```
The change drops the inline comment explaining (mostly, it ends with an unspecified core version) why this madness seems needed.

ndeloof replied:
I felt it was more confusing than helping in understanding what this is all about - at least it didn't helped me. I'm still not sure why this is needed.

ndeloof replied:
```
// collapse the structure to remain backward compatible with the JSON structure before 1.
``
`
this comment was introduced by https://github.com/jenkinsci/jenkins/commit/4ab6fe7cb4ccf2277da970447c50eab7c0921e47  (jenkins-1.494) but still doesn't make any sense to me with this context information.

Nothing get "collapsed" here afaict and this method just assumes target descriptor's json is nested within a larger json payload, anything else ? Maybe this comment was copy/pasted from another source ?

daniel-beck replied:
Well, without any comment you look at it and wonder why this mutates the passed object etc. Seems pretty insane. So restoring the extended comment and pointing to the commit perhaps would be useful?

ndeloof commented:
code review comments have been addressed, can we get this merged ?

ndeloof commented:
A major design issue I see here is that we don't give any solution to "reset to default before binding" issue, i.e is the json payload miss a value, the setter never will be executed (even with `null`) to reset field and give setter a change to use some default value.
https://github.com/stapler/stapler/pull/147 offers a possible alternative, but from the scope of this PR I can't find any reasonable way to ensure all setter get called just using the `bindJSON` approach.
Any suggestion ?

daniel-beck commented:
JENKINS-21017 all over :(

ndeloof commented:
this reminds me 2000's when I had to handle this same issue with Struts 1.0 forms :-\

ndeloof commented:
I've added support for `bindJSON(bean,json)` in https://github.com/stapler/stapler/pull/147
This enforce all DataBoundSetters in a descriptor will get invoked either with data from json payload or with default value. Doing so we are guaranteed Descriptor will be fully (re)configured

ndeloof commented:
I just noticed it might be nice we can _rollback_ changes when some FormException is thrown during data-binding. BulkChange doesn't offer this option, but I had in mind we could catch unexpected exception and invoke `load()` to restore Descriptor based on last successfully saved configuration. wdyt ?

jglick commented on core/src/main/java/jenkins/model/GlobalConfiguration.java:
```diff
@@ -57,17 +57,6 @@ public String getGlobalConfigPage() {
-     * appropriate when your implementation has getters and setters for all fields.
-     * <p>{@inheritDoc}
-     */
-    @Override
```
Thus generalizing #2509 FTR.

ndeloof committed:
new snapshot to include 9407c09

Signed-off-by: Nicolas De Loof <nicolas.deloof@gmail.com>

ndeloof commented:
`Could not find artifact org.kohsuke.stapler:stapler:jar:256-20181014.202408-2 in azure`
bu it is well deployed https://repo.jenkins-ci.org/snapshots/org/kohsuke/stapler/stapler/1.256-SNAPSHOT/stapler-1.256-20181014.202408-2.jar

ndeloof commented:
testDeleteRecursive_onWindows FAILED
`java.lang.AssertionError: deletion stopped`

not sure about the cause and if/how this is related to my PR ...

ndeloof committed:
make PersistedList a Replaceable so stapler can bind them as fields

Signed-off-by: Nicolas De Loof <nicolas.deloof@gmail.com>

ndeloof commented:
@oleg-nenashev @jglick @batmat @daniel-beck you self-assigned for review, but didn't got much feedback then. How can we process forward on this ?

jglick commented:
Indeed I have not had time to think about this.

ndeloof commented on core/src/main/java/hudson/model/Descriptor.java:
```diff
@@ -807,8 +809,33 @@ public boolean configure( StaplerRequest req ) throws FormException {
+            req.bindJSON(this, json);
+            bc.commit();
+        } catch (IOException e) {
+            LOGGER.log(Level.WARNING, "Failed to save "+getConfigFile(),e);
```
I'm a bit worried we hide failure to save configuration here. Should throw a FormException. wdyt ?

ndeloof replied:
Agree, but FormException expect a specific field, and here the whole binding process did fail. Maybe better to throw an IOException ?
No concrete sample I can find in code to see the conventional way to address this

ndeloof replied:
Actually (sorry, this PR being 1 month old I can't remember such details) the original code forhudson.model.Descriptor#save doesn't send any feedback to end-user but just log failure to persist configuration, so I just did the same here.

alecharp reviewed (commented):
I might not see the entire picture, but I couldn't see anything dramatic here.

alecharp commented on core/pom.xml:
```diff
@@ -634,6 +635,14 @@ THE SOFTWARE.
+      <groupId>org.kohsuke.stapler</groupId>
+      <artifactId>stapler</artifactId>
+      <version>1.256-20181020.094214-5</version>
+      <!-- see https://github.com/stapler/stapler/pull/147 -->
```
Perfect!

oleg-nenashev commented:
Please merge the PR with upstream to rerun the tests

oleg-nenashev reviewed (commented):
Also, IMHO it needs JIRA issue or link to any other document (like JEP) describing the change and dependencies on other stories

oleg-nenashev commented:
https://github.com/stapler/stapler/pull/147 is still needed to merge that

ndeloof commented:
created https://issues.jenkins-ci.org/browse/JENKINS-54950 for this specific PR

jeffret-b commented:
I'm going to re-open to kick a build and see if the prior failure was just an infra issue.

Wadeck commented on core/src/main/java/hudson/Functions.java:
```diff
@@ -1884,6 +1885,30 @@ public String getPasswordValue(Object o) {
+     * Retrieve default value to be used when setting field on type.
+     * Used by {@code <f:textbox>} to retrieve default value defined by component meta-data.
+     */
+    public String getDefaultValue(Class type, String field) {
```
Could you implement some unit tests for this method, especially cases for the catch statement due to the complexity ?

Wadeck commented on core/src/main/java/hudson/model/Descriptor.java:
```diff
@@ -807,8 +809,33 @@ public boolean configure( StaplerRequest req ) throws FormException {
-        // compatibility
-        return configure(req);
+
+        if (Util.isOverridden(Descriptor.class, getClass(), "configure", StaplerRequest.class)) {
```
A test case to ensure this behavior is kept seems useful, WDYT ?

Wadeck commented on core/src/main/java/hudson/Functions.java:
```diff
@@ -1884,6 +1885,30 @@ public String getPasswordValue(Object o) {
 
+    /**
+     * Retrieve default value to be used when setting field on type.
+     * Used by {@code <f:textbox>} to retrieve default value defined by component meta-data.
```
nit: + `f:number` + `f:textarea`

Wadeck reviewed (commented):
Thank you @ndeloof for the concept, it's really interesting!

I will feel more confident with more automatic tests but also at least someone that tested it manually to ensure there is no other parts that are broken (or from plugins).

ndeloof commented:
I'm closing this PR as I don't plan to invest more time on this, after months waiting for approval. Feel free to cherry-pick code if this topic is some way resurrected.
